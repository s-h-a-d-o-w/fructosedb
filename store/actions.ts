import {SupportedLanguages} from 'types';

export const changeFilter = (value: string) =>
	({type: 'CHANGE_FILTER', value} as const);
export const changeLanguage = (value: SupportedLanguages) =>
	({type: 'CHANGE_LANGUAGE', value} as const);
export const changeSort = (sortBy: string, sortAsc: boolean) =>
	({
		type: 'CHANGE_SORT',
		sortBy,
		sortAsc,
	} as const);
export const changeTranslationTarget = (value: string) =>
	({
		type: 'CHANGE_TRANSLATION_TARGET',
		value,
	} as const);
export const hideFloat = () =>
	({
		type: 'HIDE_FLOAT',
	} as const);
export const rehydrate = () =>
	({
		type: 'REHYDRATE',
	} as const);
export const showFloat = (content: string, x: number, y: number) =>
	({
		type: 'SHOW_FLOAT',
		content,
		x,
		y,
	} as const);
export const toggleFruit = () =>
	({
		type: 'TOGGLE_FRUIT',
	} as const);
export const toggleLockAvoid = () =>
	({
		type: 'TOGGLE_LOCK_AVOID',
	} as const);
export const toggleServing = () =>
	({
		type: 'TOGGLE_SERVING',
	} as const);

export type Action =
	| {type: '@@INIT'}
	| ReturnType<typeof changeFilter>
	| ReturnType<typeof changeLanguage>
	| ReturnType<typeof changeSort>
	| ReturnType<typeof changeTranslationTarget>
	| ReturnType<typeof hideFloat>
	| ReturnType<typeof rehydrate>
	| ReturnType<typeof showFloat>
	| ReturnType<typeof toggleFruit>
	| ReturnType<typeof toggleLockAvoid>
	| ReturnType<typeof toggleServing>;
