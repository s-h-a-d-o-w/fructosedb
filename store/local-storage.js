// see: https://egghead.io/lessons/javascript-redux-persisting-the-state-to-the-local-storage
import throttle from 'lodash/throttle';

export const loadState = (defaultState) => {
	try {
		return localStorage.getItem('state') === null
			? defaultState
			: JSON.parse(localStorage.getItem('state'));
	} catch (e) {
		console.error(e);
		return defaultState;
	}
};

export const saveState = throttle((state) => {
	try {
		localStorage.setItem('state', JSON.stringify(state));
	} catch (e) {
		console.error(e);
	}
}, 1000);
