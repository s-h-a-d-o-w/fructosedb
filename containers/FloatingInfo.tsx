import * as React from 'react';
import {connect} from 'react-redux';
import styled from 'styled-components';

import {ReduxState} from 'store';

export const VERTICAL_OFFSET = 30;

const StyledFloat = styled.div.attrs<Props>({
	// This prevents classes from being created with every change
	style: (props: Props) => ({
		left: `${props.float ? props.float.x : 0}px`,
		top: `${props.float ? props.float.y - VERTICAL_OFFSET : 0}px`,
	}),
})`
	font-family: 'Roboto Condensed', sans-serif;

	position: absolute;
	z-index: 10000;
	max-width: 30vw;
	padding: 0.5vw;

	background-color: white;
	color: black;

	pointer-events: none;
`;

type Props = ReturnType<typeof mapStateToProps>;

// Always rendered (not as much adding/removing from DOM, more concise code),
// won't be visible if content is empty
const FloatingInfo: React.FC<Props> = (props) =>
	props.float ? (
		<StyledFloat {...props}>{props.float.content}</StyledFloat>
	) : null;

const mapStateToProps = ({float}: ReduxState) => ({
	float,
});

export default connect(mapStateToProps)(FloatingInfo);
