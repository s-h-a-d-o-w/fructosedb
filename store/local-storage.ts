// Based on: https://egghead.io/lessons/javascript-redux-persisting-the-state-to-the-local-storage
import throttle from 'lodash/throttle';
import {ReduxState} from 'store';

export const loadState = (defaultState: ReduxState): ReduxState => {
	const storedState = localStorage.getItem('state');

	return storedState === null
		? defaultState
		: {...defaultState, ...JSON.parse(storedState)};
};

export const saveState = throttle((state) => {
	try {
		localStorage.setItem('state', JSON.stringify(state));
	} catch (e) {
		console.error(e);
	}
}, 1000);
