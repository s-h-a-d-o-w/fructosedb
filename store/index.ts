import {TypedUseSelectorHook, useSelector} from 'react-redux';
import {createStore, applyMiddleware} from 'redux';
import * as Redux from 'redux';
import {composeWithDevTools} from 'redux-devtools-extension';
import thunkMiddleware from 'redux-thunk';

import {Action} from './actions';
import {loadState, saveState} from './local-storage';
import {SupportedLanguages} from 'types';

export const initialState = {
  count: 0,
  filter: '',
  lastUpdate: 0,
  lang: 'en' as SupportedLanguages,
  langTranslate: 'en',
  light: false,
  onlyFruit: false,
  showServingSize: false,
  sortBy: 'name',
  sortAsc: true,
};

export const reducer = (
  state: ReduxState = initialState,
  action: Action
): ReduxState => {
  switch (action.type) {
    case 'CHANGE_FILTER':
      return {
        ...state,
        filter: action.value,
      };
    case 'CHANGE_LANGUAGE':
      return {
        ...state,
        lang: action.value,
      };
    case 'CHANGE_SORT':
      return {
        ...state,
        sortBy: action.sortBy,
        sortAsc: action.sortAsc,
      };
    case 'CHANGE_TRANSLATION_TARGET':
      return {
        ...state,
        langTranslate: action.value,
      };
    case 'HIDE_FLOAT':
      if (!state.float) return state;

      const {float, ...stateWithoutFloat} = state;
      return stateWithoutFloat;
    case 'REHYDRATE':
      return loadState(state);
    case 'SHOW_FLOAT':
      return {
        ...state,
        float: {
          content: action.content,
          x: action.x,
          y: action.y,
        },
      };
    case 'TOGGLE_FRUIT':
      return {
        ...state,
        onlyFruit: !state.onlyFruit,
      };
    case 'TOGGLE_SERVING':
      return {
        ...state,
        showServingSize: !state.showServingSize,
      };
    default:
      return state;
  }
};

// INITIALIZATION
export type ReduxState = typeof initialState & {
  float?: {
    x: number;
    y: number;
    content: string;
  };
};

// Locale is needed so that we can set the default language to
// match the SSRed content.
export const initializeStore = (
  locale: string | null,
  state: object = initialState
): Redux.Store => {
  const store = createStore(
    reducer,
    {...state, lang: locale ? (locale as SupportedLanguages) : 'en'},
    composeWithDevTools(applyMiddleware(thunkMiddleware))
  );

  if (typeof window !== 'undefined') {
    store.subscribe(() => saveState(store.getState()));
  }

  return store;
};

export const useTypedSelector: TypedUseSelectorHook<ReduxState> = useSelector;
