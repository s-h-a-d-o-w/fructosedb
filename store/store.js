import {createStore, applyMiddleware} from 'redux';
import {composeWithDevTools} from 'redux-devtools-extension';
import thunkMiddleware from 'redux-thunk';

import {loadState, saveState} from './local-storage.js';
import {isEmptyObject} from '../lib/util';

export const actionTypes = {
	CHANGE_FILTER: 'CHANGE_FILTER',
	CHANGE_LANGUAGE: 'CHANGE_LANGUAGE',
	CHANGE_SORT: 'CHANGE_SORT',
	CHANGE_TRANSLATION_TARGET: 'CHANGE_TRANSLATION_TARGET',

	KILL_FLOAT: 'KILL_FLOAT',
	REHYDRATE: 'REHYDRATE',
	SHOW_FLOAT: 'SHOW_FLOAT',

	TOGGLE_FRUIT: 'TOGGLE_FRUIT',
	TOGGLE_LOCK_AVOID: 'TOGGLE_LOCK_AVOID',
	TOGGLE_SERVING: 'TOGGLE_SERVING',
};

export const reducer = (state = defaultInitialState, action) => {
	switch (action.type) {
		case actionTypes.CHANGE_FILTER:
			return Object.assign({}, state, {filter: action.value});
		case actionTypes.CHANGE_LANGUAGE:
			return Object.assign({}, state, {lang: action.value});
		case actionTypes.CHANGE_SORT:
			return Object.assign({}, state, {
				sortBy: action.sortBy,
				sortAsc: action.sortAsc,
			});
		case actionTypes.CHANGE_TRANSLATION_TARGET:
			return Object.assign({}, state, {langTranslate: action.value});
		case actionTypes.KILL_FLOAT:
			return isEmptyObject(state.float)
				? state
				: Object.assign({}, state, {
						float: {},
				  });
		case actionTypes.REHYDRATE:
			return loadState(state);
		case actionTypes.SHOW_FLOAT:
			return Object.assign({}, state, {
				float: {
					content: action.content,
					x: action.x,
					y: action.y,
				},
			});
		case actionTypes.TOGGLE_LOCK_AVOID:
			return Object.assign({}, state, {
				lockedAvoid: !state.lockedAvoid,
			});
		case actionTypes.TOGGLE_FRUIT:
			return Object.assign({}, state, {
				onlyFruit: !state.onlyFruit,
			});
		case actionTypes.TOGGLE_SERVING:
			return Object.assign({}, state, {
				showServing: !state.showServing,
			});
		default:
			return state;
	}
};

export const actions = {
	changeFilter: (value) => ({type: actionTypes.CHANGE_FILTER, value}),
	changeLanguage: (value) => ({type: actionTypes.CHANGE_LANGUAGE, value}),
	changeSort: (sortBy, sortAsc) => ({
		type: actionTypes.CHANGE_SORT,
		sortBy,
		sortAsc,
	}),
	changeTranslationTarget: (value) => ({
		type: actionTypes.CHANGE_TRANSLATION_TARGET,
		value,
	}),
	killFloat: () => ({
		type: actionTypes.KILL_FLOAT,
	}),
	rehydrate: () => ({
		type: actionTypes.REHYDRATE,
	}),
	showFloat: (content, x, y) => ({
		type: actionTypes.SHOW_FLOAT,
		content,
		x,
		y,
	}),
	toggleFruit: () => ({
		type: actionTypes.TOGGLE_FRUIT,
	}),
	toggleServing: () => ({
		type: actionTypes.TOGGLE_SERVING,
	}),
};

// INITIALIZATION
const defaultInitialState = {
	count: 0,
	float: {},
	lastUpdate: 0,
	lang: 'en',
	light: false,
	lockedAvoid: true,
	onlyFruit: false,
	showServing: false,
	sortBy: 'name',
	sortAsc: true,
};

export const initializeStore = (initialState = defaultInitialState) => {
	return createStore(
		// No persistence on the server.
		typeof window === 'undefined'
			? reducer
			: (...args) => {
					let state = reducer.apply(null, args);

					// Need to skip @@INIT, since that would obviously overwrite whatever state there
					// might be in local storage before we can even use it for rehydration.
					if (args[1].type !== '@@INIT') saveState(state);

					return state;
			  },
		initialState,
		composeWithDevTools(applyMiddleware(thunkMiddleware))
	);
};
