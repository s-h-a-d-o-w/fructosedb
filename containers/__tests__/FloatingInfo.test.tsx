import React from 'react';
import {mount} from 'enzyme';
import {render} from '@testing-library/react';
import * as Redux from 'redux';
import {Provider} from 'react-redux';
import 'jest-styled-components';

import {initialState, initializeStore} from 'store';

import FloatingInfo, {VERTICAL_OFFSET} from '../FloatingInfo';

let store: Redux.Store;

describe('<FloatingInfo>', () => {
	beforeEach(() => {
		store = initializeStore(
			Object.assign({}, initialState, {
				float: {
					x: 100,
					y: 100,
					content: 'unit test',
				},
			})
		);
	});

	it('snapshots correctly', () => {
		const {container} = render(
			<Provider store={store}>
				<FloatingInfo />
			</Provider>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it('sets "top" to the correct offset', () => {
		const wrapper = mount(
			<Provider store={store}>
				<FloatingInfo />
			</Provider>
		);
		expect((wrapper.find('div').prop('style') as CSSStyleDeclaration).top).toBe(
			`${100 - VERTICAL_OFFSET}px`
		);
	});

	it("doesn't render if float is empty", () => {
		store = initializeStore(initialState);

		const wrapper = mount(
			<Provider store={store}>
				<FloatingInfo />
			</Provider>
		);

		expect(wrapper.find('div').length).toEqual(0);
	});
});
