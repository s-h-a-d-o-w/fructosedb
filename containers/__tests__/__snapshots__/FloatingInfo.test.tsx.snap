// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FloatingInfo> snapshots correctly 1`] = `
.c0 {
  font-family: 'Roboto Condensed',sans-serif;
  position: absolute;
  z-index: 10000;
  max-width: 30vw;
  padding: 0.5vw;
  background-color: white;
  color: black;
  pointer-events: none;
}

<div
  class="c0"
  style="left: 100px; top: 70px;"
>
  unit test
</div>
`;
